File: BitMasking/1182.py
import sys

N,S = map(int,sys.stdin.readline().split())
li = list(map(int,sys.stdin.readline().split()))
bits = list(0 for i in range(N))
ans = 0
def recur(idx):
    global bits, li, ans
    if idx == N:
        tmp = 0
        cnt = 0
        for i in range(len(bits)):
            if bits[i]:
                tmp += li[i]
                cnt += 1
        if tmp == S and cnt>0:
            ans += 1
        return
    recur(idx+1)
    bits[idx] = ~bits[idx]
    recur(idx+1)

recur(0)
print(ans)

File: BruteForce/1182.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N,S, ans=0;
    static int[] arr;
    public static void main(String[] args) throws IOException{
        int[] tmp = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        N = tmp[0]; S = tmp[1];
        arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        for(int i=0;i<N;i++)
            recur(i,arr[i]);
        System.out.println(ans);
    }
    public static void recur(int idx, int sum){
        if(sum == S)
            ans++;
        for(int i=idx+1;i<N;i++){
            recur(i,sum+arr[i]);
        }
    }
}